File: plugins/plugin-mattermost/typescript/src/providers/chatState.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";
import { getChannelKind } from "../types";

export const CHAT_STATE_PROVIDER = "mattermost_chat_state";

export const chatStateProvider: Provider = {
  name: CHAT_STATE_PROVIDER,
  description:
    "Provides Mattermost chat context including channel ID, user ID, team ID, and channel type",
  dynamic: true,

  get: async (_runtime: IAgentRuntime, message: Memory, _state: State): Promise<ProviderResult> => {
    const channelId = message.content?.channelId as string | undefined;
    const userId = message.content?.userId as string | undefined;
    const metadata = message.content?.metadata as Record<string, unknown> | undefined;
    const postId = metadata?.postId as string | undefined;
    const rootId = metadata?.rootId as string | undefined;
    const teamId = metadata?.teamId as string | undefined;
    const channelType = metadata?.channelType as string | undefined;
    const roomId = message.roomId;

    const kind = getChannelKind(channelType);
    const isDm = kind === "dm";
    const isGroup = kind === "group";
    const isChannel = kind === "channel";
    const isThread = Boolean(rootId);

    const data = {
      channelId,
      userId,
      postId,
      rootId,
      teamId,
      channelType,
      roomId,
      isDm,
      isGroup,
      isChannel,
      isThread,
    };

    const values: Record<string, string> = {
      channel_id: channelId || "",
      user_id: userId || "",
      post_id: postId || "",
      root_id: rootId || "",
      team_id: teamId || "",
      channel_type: channelType || "",
      room_id: roomId || "",
      is_dm: isDm.toString(),
      is_group: isGroup.toString(),
      is_channel: isChannel.toString(),
      is_thread: isThread.toString(),
    };

    let text = "Mattermost Chat State:\n";
    if (channelId) {
      text += `Channel ID: ${channelId}\n`;
    }
    if (channelType) {
      const typeLabel = isDm ? "Direct Message" : isGroup ? "Group Message" : "Channel";
      text += `Channel Type: ${typeLabel}\n`;
    }
    if (userId) {
      text += `User ID: ${userId}\n`;
    }
    if (teamId) {
      text += `Team ID: ${teamId}\n`;
    }
    if (rootId) {
      text += `Thread Root: ${rootId}\n`;
    }

    return {
      data,
      values,
      text,
    };
  },
};

File: plugins/plugin-mattermost/rust/src/providers/chat_state.rs
use async_trait::async_trait;
use serde_json::Value;

use super::{MattermostProvider, ProviderContext};
use crate::types::MattermostChannelType;

/// Provider that exposes the current chat/channel context as JSON.
pub struct ChatStateProvider;

#[async_trait]
impl MattermostProvider for ChatStateProvider {
    fn name(&self) -> &'static str {
        "mattermost_chat_state"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        let channel_type = context
            .channel_type
            .as_ref()
            .and_then(|t| MattermostChannelType::from_str(t));

        let is_dm = channel_type
            .map(|t| matches!(t, MattermostChannelType::Direct))
            .unwrap_or(false);
        let is_group = channel_type
            .map(|t| matches!(t, MattermostChannelType::Group))
            .unwrap_or(false);
        let is_channel = channel_type
            .map(|t| {
                matches!(
                    t,
                    MattermostChannelType::Open | MattermostChannelType::Private
                )
            })
            .unwrap_or(false);
        let is_thread = context.root_id.is_some();

        serde_json::json!({
            "channel_id": context.channel_id,
            "user_id": context.user_id,
            "post_id": context.post_id,
            "root_id": context.root_id,
            "team_id": context.team_id,
            "channel_type": context.channel_type,
            "room_id": context.room_id,
            "is_dm": is_dm,
            "is_group": is_group,
            "is_channel": is_channel,
            "is_thread": is_thread
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_chat_state_dm() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            channel_id: Some("channel123".to_string()),
            user_id: Some("user456".to_string()),
            post_id: Some("post789".to_string()),
            root_id: None,
            team_id: Some("team111".to_string()),
            channel_type: Some("D".to_string()),
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["channel_id"], "channel123");
        assert_eq!(state["is_dm"], true);
        assert_eq!(state["is_group"], false);
        assert_eq!(state["is_thread"], false);
    }

    #[tokio::test]
    async fn test_chat_state_thread() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            channel_id: Some("channel123".to_string()),
            user_id: Some("user456".to_string()),
            post_id: Some("post789".to_string()),
            root_id: Some("root111".to_string()),
            team_id: Some("team111".to_string()),
            channel_type: Some("O".to_string()),
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["is_dm"], false);
        assert_eq!(state["is_channel"], true);
        assert_eq!(state["is_thread"], true);
    }
}
